File: season2/2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        carry = 0
        head = curr = ListNode()
        
        while l1 or l2:
            total = carry
            if l1:
                total += l1.val
                l1 = l1.next
            if l2:
                total += l2.val
                l2 = l2.next
            
            carry = total // 10
            if carry:
                total = total % 10
            
            curr.next = ListNode(total)
            curr = curr.next
        
        if carry:
            curr.next = ListNode(carry)
            curr = curr.next
        
        return head.next
            
            
            
            
                
        

File: season2/2.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function (l1, l2) {
  let carry = 0,
    sum = 0,
    dummy = new ListNode();
  let curr = dummy;
  while (l1 || l2) {
    //console.log(carry, l1.val, l2 && l2.val);
    sum = 0;
    sum += carry;
    sum += l1 ? l1.val : 0;
    sum += l2 ? l2.val : 0;

    carry = 0;

    if (sum >= 10) {
      //console.log('sum', sum);
      carry = Math.floor(sum / 10);
      sum = sum % 10;
      //console.log(carry, sum);
    }

    curr.next = new ListNode(sum);
    curr = curr.next;

    if (l1) l1 = l1.next;
    if (l2) l2 = l2.next;
  }

  if (carry) curr.next = new ListNode(carry);

  return dummy.next;
};
